caravel/models.py

```diff
@@ -1974,6 +2001,14 @@ def recursive_get_fields(_conf):
         cols += [col for col in groupby if col in df.columns]
         cols += [col for col in metrics if col in df.columns]
         df = df[cols]
+
+        time_offset = DruidDatasource.time_offset(granularity)
+        if 'timestamp' in df.columns and time_offset:
+            for i in range(len(df['timestamp'])):
```

Comment:
oh with pandas dataframe you always want to do set operations, something like `df.timestamp += timedelta(days=1)` should work. If you're doing it right, there's almost never a need to write a loop.